### PR TITLE
BUG: fix reference count leak in __array__ internals

### LINKED ISSUE #26020: BUG: NumPy 2.0.0b1 leaks reference counts in `np.array`
### Describe the issue:

`np.array` in NumPy 2.0.0b1 can leak reference counts to its inputs, potentially causing memory leaks. This behavior is not observed in NumPy 1.X releases

### Reproduce the code example:

```python
import sys
import gc
import numpy as np

print(np.__version__)

class NotAnArray:
    def __array__(self):
        raise NotImplementedError()
x = NotAnArray()

print(sys.getrefcount(x))
try:
    np.array(x)
except:
    pass
print(sys.getrefcount(x))
gc.collect()
print(sys.getrefcount(x))
```

### Error message:

Output in NumPy 2.0.0b1:
```
2.0.0b1
2
3
3
```
Output in NumPy 1.26.4:
```
1.26.4
2
2
2
```

### Python and NumPy Versions:

```
>>> import sys, numpy; print(numpy.__version__); print(sys.version)
2.0.0b1
3.10.12 (main, Nov 20 2023, 15:14:05) [GCC 11.4.0]
```

### Runtime Environment:

```
>>> import numpy; print(numpy.show_runtime())
[{'numpy_version': '2.0.0b1',
  'python': '3.10.12 (main, Nov 20 2023, 15:14:05) [GCC 11.4.0]',
  'uname': uname_result(system='Linux', node='6ccda8d672a0', release='6.1.58+', version='#1 SMP PREEMPT_DYNAMIC Sat Nov 18 15:31:17 UTC 2023', machine='x86_64')},
 {'simd_extensions': {'baseline': ['SSE', 'SSE2', 'SSE3'],
                      'found': ['SSSE3',
                                'SSE41',
                                'POPCNT',
                                'SSE42',
                                'AVX',
                                'F16C',
                                'FMA3',
                                'AVX2'],
                      'not_found': ['AVX512F',
                                    'AVX512CD',
                                    'AVX512_KNL',
                                    'AVX512_KNM',
                                    'AVX512_SKX',
                                    'AVX512_CLX',
                                    'AVX512_CNL',
                                    'AVX512_ICL']}}]
```

### Context for the issue:

This manifested as a JAX tracer leak and was caught by JAX upstream tests; see https://github.com/google/jax/issues/20254

### PR DESCRIPTION
**Pull request by ngoldbaum**
Closes #26020.

Would very much appreciate another careful eye on this or suggestions to make the error handling less convoluted.

### CODE REVIEW

File: numpy/_core/src/multiarray/ctors.c [hunk 1]
```diff
@@ -2466,6 +2466,7 @@ PyArray_FromArrayAttr_int(
 
     if (new == NULL) {
         if (npy_ma_str_array_err_msg_substr == NULL) {
+            Py_DECREF(array_meth);
```
Comment (jakevdp) [inline numpy/_core/src/multiarray/ctors.c:2469]: Do `args` and `kwargs` need to be DECREF'd here?

File: numpy/_core/tests/test_multiarray.py [hunk 2]
```diff
@@ -8492,6 +8492,23 @@ def __array__(self, dtype=None):
...
+
+        x = NotAnArray()
+
+        refcount = sys.getrefcount(x)
+
+        try:
+            np.array(x)
+        except NotImplementedError:
+            pass
+
```
Comment (jakevdp) [inline numpy/_core/tests/test_multiarray.py:8509]: Do we need `gc.collect()` here to prevent test flakiness?

**Review by jakevdp — COMMENTED** [review]
Thanks for looking into this!

**Review by mhvk — COMMENTED** [review]
Looks good modulo the two points raised by Jake

File: numpy/_core/tests/test_multiarray.py (hunk 2 above)
Comment (ngoldbaum) [reply numpy/_core/tests/test_multiarray.py:8509]: It doesn't look like we do that elsewhere in this file but can't hurt.

File: numpy/_core/src/multiarray/ctors.c (hunk 1 above)
Comment (ngoldbaum) [reply numpy/_core/src/multiarray/ctors.c:2469]: Thanks for spotting

File: numpy/_core/tests/test_multiarray.py (hunk 2 above)
Comment (seberg) [reply numpy/_core/tests/test_multiarray.py:8509]: There are no reference cycles possible here, so I don't think it matters either way.

**Review by seberg — APPROVED** [review]
Would be much better with a `goto finish` pattern (and I still think using vectorcall would both be faster and simpler) but lets start with this to get the leak out.

**Pull request merged**